### ISSUE TITLE
[Proposal]: Only Allow Lexical Keywords in the Language

### ISSUE THREAD

**Issue by jmarolf** [dotnet/csharplang#4460]
# Only Allow Lexical Keywords in the Language

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Today there are keywords in the language that cannot be understood with just lexical information such as `var` or `nameof`. These keywords operate this way for backwards compatibility reasons. I propose that we change the language so that there exist no keywords that cannot be determined via lexical analysis. 

## Motivation
[motivation]: #motivation

In general, C# strives to be a language that is explicit about program behavior and normally requires the developer to write out what their intention is without ambiguities. This makes it a language that is easy for someone to read and understand. Once you learn it there is little implicit behavior to consider. I believe having this "gotcha" where keywords are only keywords if nothing else in scope is so-named makes the language harder to read and reason about in general.

There is also the reality that the design goals around C# Language versions and .NET Framework versions have changed. In the past it was paramount that developers could take a new language version update without updating the framework version they were targeting. With language features being increasingly tied to the runtime this makes less sense. We now strongly encourage developers to update both the language version and target framework together.

## Detailed design
[design]: #detailed-design

There is an existing concept in the language called "contextual keywords". I am not proposing doing away with this concept altogether just changing it so that a keyword's "contextual-ness" is always able to be determined lexically. Take new (at the time of this writing) keyword `record`. We can still know if we are referring to the `record` keyword or some identifier named record based on the lexical context, there is no ambiguity. However `var`, according to the spec, requires us to check if there are types named var in scope:

[spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/statements#local-variable-declarations)
> In the context of a local variable declaration, the identifier var acts as a contextual keyword. When the local_variable_type is specified as var and no type named var is in scope,

Similarly, `nameof` requires checking if a there are any identifiers called nameof in scope

[spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/statements#local-variable-declarations)

> Because nameof is not a reserved keyword, a nameof expression is always syntactically ambiguous with an invocation of the simple name nameof. For compatibility reasons, if a name lookup of the name nameof succeeds, the expression is treated as an invocation_expression -- regardless of whether the invocation is legal. Otherwise it is a nameof_expression.

The implementation of this proposal would remove wording from the spec around name lookup collisions, and have a compliant compiler be able to fully determine keywords given only parsing information.

The following keywords would now error if developers attempted to use them as anything other than a keyword
- `var`
- `nameof`
- `dynamic`
- `_`

## Drawbacks
[drawbacks]: #drawbacks

This is a breaking change, if anyone were relying on this behavior in their code it would no longer compile. For cases where a type is named `var`, `dynamic`, or `_` or a method is called `nameof` the developer would need to change the usages to `@var`, `@nameof`, `@dynamic`, or `@_`. 

## Alternatives
[alternatives]: #alternatives

We could opt to keep `_` as a contextual keyword that depends on name lookup rules as this is the change most likely to break real-world programs (see discussion on https://github.com/dotnet/csharplang/issues/1064)

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by jaredpar** [comment]
>  (see discussion on #1064)

Did you mean to have this refer to the discussion, not this same issue?

**Comment by svick** [comment]
Previous discussion: https://github.com/dotnet/csharplang/discussions/4458.

**Comment by svick** [comment]
I feel that the breaking changes that would be introduced by adopting this proposal on its own fall into three categories:

1. "Nobody is broken."
    Almost nobody uses `var`, `dynamic` or `nameof` as identifiers in C#, it's fine to break the tiny number of people who do.
2. "Some people are broken."
    The pattern where `_` is the name of a used lambda parameter (e.g. `_ => _.Name`) is fairly rare, but not unheard of in C#. I think it's probably acceptable to break this kind of code, but ways of softening the blow should be seriously considered (e.g. a code fix to rename such parameters; or warning in C# 10 and only making it an error in C# 11).
3. "Lots of people are broken."
    The pattern where `_` is the name of an unused lambda parameter (e.g. `_ => {}`) is very common in C# and it is completely unacceptable to break such code. The obvious solution would be to change the meaning of that code to make the `_` a discard. But I think it's important to note that this additional change would be required, assuming this proposal is not meant to be a massive breaking change.

**Comment by Youssef1313** [comment]
- Will this break analyzers/codefixes?

**Comment by jmarolf** [comment]
@svick 
looking at github we have:
 - [~17K](https://grep.app/search?q=_%20%3D%3E&filter[lang][0]=C%23) uses of `_ =>`
 - [~1K](https://grep.app/search?q=var%20_%20%3D&filter[lang][0]=C%23) uses of `var _ =` or `using var _ =`

While I am only proposing removing name lookup and `var _ = GetResults();` is not lexically ambiguous with `_ = GetResults();` there could potentially be odd errors with `_` in this case. I am willing to say we keep the name lookup rules for `_` if there are concerns.

**Comment by jmarolf** [comment]
> Will this break analyzers/codefixes?

that is entirely dependent on the compiler implementation, but new language versions are always allowed to break  analyzers.

Consider when expression-bodied-members were added. If you previously assumed (not unreasonably) in your analyzer that all methods with a body contained a block syntax you were broken since now the body could just be an arrow expression.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/4466

Conflating the two was always going to create a lot of confusion, and the parser that get's broken/confused the most is the human parser.

**Comment by jnm2** [comment]
I've switched my position on https://github.com/dotnet/csharplang/issues/1064 (disallowing `_` as an identifier) from downvote to upvote. Sure, it would break past years of my code when I used this for lambda parameters, but ship a solution-wide light bulb fix for it and I'd want to use that anyway to replace my usages of `_`, even if I wasn't forced to. Opting into a new major versions of C# feels like an expected time for something like this to happen.

Maybe soft-deprecating by adding a new compiler warning in C# 10 that tells you that `_` as an identifier will be disallowed starting in C# 11 would make this seem less abrupt.

I agree with @HaloFour. I think the human parser should be the most important factor. While unlikely, the possibility of code like this should make us uneasy:

```cs
public class Foo
{
    private int _;

    // Doing something important in some other file that is affected by reading Foo.WrappedValue?
    public int WrappedValue => _;

    public bool IsNumber(string input)
    {
        return double.TryParse(input, out _); // oops!
    }
}
```

(stolen from http://gafter.blogspot.com/2017/06/making-new-language-features-stand-out.html?showComment=1509474504510#c7458806139970524286)

**Comment by chrizpro** [comment]
I hope this proposal goes nowhere. I like the underscore usage because it makes the code less boring. As a writer, I like to use dashes, colons, semicolons, and etc to make my writing more interesting just like an _ makes the code more interesting, though it's rarely used. I want C# to be an intermediate-level language and developers having trouble with var and etc, they should look at Lua or etc.

**Comment by HaloFour** [comment]
Java, which is used by a larger (and some would argue more resistant to change) community seemingly has had little/no issue with the language deprecating and then disallowing the use of `_` as an identifier or `var` as a type name.  They've done this in the past as well with names like `assert`.  Usually it doesn't matter as the contextual keyword wouldn't be expected to be used as a type name.

I've been pretty outspoken against the use of `_` as a discard as well as an identifier.  Since the ship has sailed on discards I think its use as an identifier should be reconsidered.  It sounds good on paper to avoid reinterpreting/breaking any existing code, but now the language has this wart where developers need to remember which combination of features will cause the compiler to prefer `_` as an identifier vs. where the compiler will always consider it a discard, and where `_` is preferred to be an identifier the developer has to wade past the type checks on these "variables" that the developer never intended to actually use.  The case of accidentally overwriting some field name might be pathological, but the mental burden on the developer will still always be there.  I would've much preferred if the compiler phased out `_` as an identifier over a few releases, with fixers to replace it with another identifier, and then switch it to a discard wholesale.  Names are cheap.  Contextual keywords that change their meaning based on nuanced use of other language features and where the contexts are very likely to collide often are not.

**Comment by jmarolf** [comment]
My views are basically the same as @HaloFour 's. I think semantically contextual keywords make a lot of sense in theory but add developer complexity and overhead for little benefit.

**Comment by CyrusNajmabadi** [comment]
> for little benefit.

I think this is very debatable.  Consider the work we're doing to support `field` inside properties in C# 10.  If we make these keywords and not contextual keywords, we simply break people (including ourselves).  And we break them *despite* them having done nothing wrong.  For example, it would break code that is totally normally and reasonable and not at all deviating from teh norms of the ecosystem at all.  I don't like the idea that someone coudl follow every best practice we gave, and then end up breaking just for expediency on our part.  In most (all?) cases, supporting semantic contextual keywords is not difficult.  Indeed, it's one of the simpler things to support.  You simply bind as normal and accept the prior meaning if it is valid.  If it isn't, then you allow the new meaning.  This means we can gently add new things to the language and not have to worry at all about breaking people.

**Comment by jmarolf** [comment]
> Consider the work we're doing to support field inside properties in C# 10.

For clarification: this proposal explicitly states that cases like this should keep working as they do, contextual keywords will always exist. Just because properties use the `value` contextual keyword does not mean that we should now force that to be a keyword _at all times_. This proposal is about ensuring that contextual keywords can always be determined based solely on lexical information as opposed to semantic information. 

> In most (all?) cases, supporting semantic contextual keywords is not difficult. Indeed, it's one of the simpler things to support. You simply bind as normal and accept the prior meaning if it is valid. If it isn't, then you allow the new meaning. This means we can gently add new things to the language and not have to worry at all about breaking people.

I totally agree that there is not engineering reason to change this. It just works for the compiler folks (as far as I am aware). But I think it add an unnecessary burden on programmers using the language. Things like `var` and `nameof` feel very unfortunate to me. Anyone following best practices in C# does not expect `var` to be unavailable to them or for `nameof` to have different semantics based on exoteric name lookup rules. It feels like a real "gotcha" moment where I can go on twitter and "well actually" anyone that uses code with these semantic contextual keywords and say "Oh you are actually not discarding that but assigning a value to a variable named `_`".

Every other language I've encounterd (C++, Jave, Typescript, Pytho, Go) does not use name lookup rules to determine whether something is a keyword (including F# and Visual Basic) and there have been no complaints. I personally feel that all this concern over keyword breakage has no real evidence, its all theoretical. Java can just add a new keyword if they need to and no one complains.

**Comment by CyrusNajmabadi** [comment]
> This proposal is about ensuring that contextual keywords can always be determined based solely on lexical information as opposed to semantic information.

Right.  but the problem with that is that it direclty goes against design goals we have for these features. for example, we want you to just be able to say `field`.  There's nothign lexical/syntactic to distinguish that this is special.  It's just going to reference the auto-prop field if nothing else binds.

> ut I think it add an unnecessary burden on programmers using the language

I don't really see this as a burden.  For people just using the language, using `var` is going to work.  So what needs to be fixed?  Same with nameof, etc.  People using our APIs could certainly be better served here with better APIs.   but that would be a roslyn concern.

> and there have been no complaints. 

This is not true.  TAke 'go' for example.  There are *lots* of complaints about the verbosity of the language.  And part of htat verbosity arises because the language doesn't want to get into this space.  So it ensures all it's constructs are extremely verbose and often unweildy, just so it doesn't have to do any semantic checks on this sort of thing.  It's a tradeoff they made, but which we're quite loathe to as it really just bulks up the language.

**Comment by jmarolf** [comment]
> Right. but the problem with that is that it direclty goes against design goals we have for these features. for example, we want you to just be able to say field. There's nothign lexical/syntactic to distinguish that this is special. It's just going to reference the auto-prop field if nothing else binds.

I would need to review the proposal but isn't this going to work exactly like `value`? You can just say that `field` is reserved now and you use `@field` if you need to "escape" the fact that this is a keyword now. I think this is an important distinction to the reader. You now have to explicitly state what your intent is. You are essentially saying "a casual reading of this might lead to believe this is the `field` keyword, which has specific semantics but that is not what is happening here, this is a custom instance and `@field` clues you into what is happening." If we were to do it all over again would we have _everything_ be a contextual keyword? I dunno I suppose I could see the argument, why put roadblocks in folks way. My position is that it's a weird language corner case that most C# developers are not aware of and is surprising to them when they learn about it.

If there is a design goal that can only be achieve with name lookup rules or everyone else in the LDM just disagrees and thinks that semantic contextual keywords are awesome and we wish we did them more often great! Thats not my position but I am willing to be convinced.

**Comment by CyrusNajmabadi** [comment]
> I would need to review the proposal but isn't this going to work exactly like value?

No. 'value' always binds to the property parameter prior to anything else in a higher scope.  `field` will not (As that would break existing, perfectly fine code).  

> You can just say that field is reserved now and you use @field if you need to "escape" the fact that this is a keyword now

That would break lots of code taht is totally fine today and which wasn't doing anything strange or inappropriate.  I do not see how customers are helped by just changing the meaning of their code on them.

> You are essentially saying "a casual reading of this

**Comment by CyrusNajmabadi** [comment]
> My position is that it's a weird language corner case that most C# developers are not aware of and is surprising to them when they learn about it.

Weird corner cases are *always* like that.  But we have tons of those *everywhere*.  The question is: is getting rid of weird corners better or worse than breaking code?  The position we've landed on generally comes down to:

1. is the code that is breaking reasonable?  or is it unreasonable?
2. is it widespread, or likely not used at all?

If it's unreasonable (which often comes down to debate) we are more likely to take the stance: trying to prop up this code is not worth it, so we would prefer to change it and accept that pathological cases break.

Similarly, if something is widespread, then we've already opened the barn door.  People clearly are using the language in this fashion in a significant fashion, and I think we have to accept that.

Where we have room to play around with is when you get into teh 'unreasonable, and not used (or very very rarely used) territory.  This is like someone coming along now and saying: yeah, i'm going to name my type `var` even though .net naming conventions (both formal and informal) from day 1 have been that types are PascalCased.  This is both unreasonable IMO for somoene to do this, and likely extraordinarily niche.  (Inded, my expectation is taht tihs only exists in projects that seek to subvert the language/compiler, in which case i don't think of that as a reasonable thing to cater to).

--

So, in the case of *some* keywords (`var`, `record`, etc.) i'm actually ok with us taking over and saying: yeah, at thsi point this is ours.  Reasonable codebases won't have any pain *at all* moving to this.

However, for some keywords, i'm not ok with us doing this.  If the pattern is eitehr reasonable, or widespread, we need to accept that and not harm users when we have a perfectly suitable way to both introduce the feature and keep things working just fine.

**Comment by jmarolf** [comment]
> So, in the case of some keywords (var, record, etc.) i'm actually ok with us taking over and saying: yeah, at thsi point this is ours. Reasonable codebases won't have any pain at all moving to this.

> However, for some keywords, i'm not ok with us doing this. If the pattern is eitehr reasonable, or widespread, we need to accept that and not harm users when we have a perfectly suitable way to both introduce the feature and keep things working just fine.

I think this is a totally reasonable stance to take. `var` feel pretty uncontroversial (to me) but other keywords feel much further along in the spectrum of causing unreasonable harm. If the LDM says "`var` should just be a keyword but these other I think should stay as they are" I would be totally fine with that. I Just want us to take the time to re-evaluate this and make sure we still feel the same way.

In the past there were more situations where a newer version of C# could be "pushed" on you. Today it's an explicit decision to update your SDK version to get an updated version of C#. Major SDK versions also have major breaking changes (api names changes etc.) to the point that developers expect some friction. I think its not unreasonable to have folks change `field` to `@field` in these upgrade situations but I will admit I am taking a stance that is way over to one side on how ok I am with breaks. Others do not need to join me over here.

**Comment by CyrusNajmabadi** [comment]
If we scope this to `the language reserves the space of lowercase ascii identifiers for **type** contexts`, then i'm totally ok with that :)

**Comment by jaredpar** [comment]
I like the scoping here but I would also like to consider the case of discards. As the feature is written today it's hard to use discards broadly in a method and instead is most useful in a limited set of circumstances. In too many cases it subtly turns into an identifier, not a discard, and suddenly that invalidates other uses within the method body and suddenly you have to drop back to ignored names.

**Comment by CyrusNajmabadi** [comment]
I'd def like to break out an issue on discards.  I;m curious about hte cases that are hard here and where it's difficult to mesh the idea of:

1. use existing semantics if the code is legal
2. reinterpret as discard if not

I think discards are also a space we could potentially experiment with a `.net upgrade` style approach where we unilaterally reinterpretted this stuff, but had tools fix the issue if you used these as non-discards in your project.

**Comment by jaredpar** [comment]
> I;m curious about hte cases that are hard here 

Converting between lambdas and local functions. Parameters in lambdas can be discards but not in local functions. That means when swapping between the two it introduces unnecessary friction because you have to rationalize discard behavior. It's no longer what essentially amounts to a syntax transform.

Whether a `_` is a discard or identifier in a lambda comes down to the count of parameters that you have. A single parameter means it's an identifier but multiple mean it's a discard. 

```c#
// _ is an identifier 
Action<int> action = (_) => { 
    _ = ""; // Error cause _ is an int identifier 
};

// _ is a discard
Action<int, int> action2 = (_, _) => { 
    _ = ""; // Okay cause this is an identifier 
};
```

This is generally frustrating to have to remember but really gets frustrating when you consider it in the context of refactoring or code changes. Consider that lambda parameters are often listed as discards because they're a callback value that you may not need. Circumstances change and it's rational to begin using a parameter which begins by assigning it a name. If assigning that parameter a name though means there is only one `_` remaining then it becomes an identifier and suddenly all the other `_` inside the method body are now interpretted as identifiers which can cause compilation errors. 

```c#
string token = ...;
Action<string, string> = (_, value) => {
     // Error: This worked before I changed the second parameter to have a name
     if (int.TryParse(token, out _) { 
        ...    
}
```

This though means there is a huge incentive to prefer `out var _` over `out _` even when `_` currently points to a discard. The `out var _` form is one of the few places where `_` unambiguously refers to a discard. Yet even though `_` is more succinct developers should consider always using the `out var _` form, even though it's longer and doesn't actually declare a variable, because it's more future proof to cases where `_` gets bound as a discard. 

These together all make it frustrating to use discards. It's too easy to get trapped in a case where `_` suddenly binds to an identifier and that will invalidate many other cases in the method where you depended on having discards available and there is little recourse for the developer when that happens.

**Comment by CyrusNajmabadi** [comment]
> Whether a _ is a discard or identifier in a lambda comes down to the count of parameters that you have. A single parameter means it's an identifier but multiple mean it's a discard.

Could we change that and instead make it so that if it's an error with the prior semantics, then it can now be reinterpretted as a discard?

> These together all make it frustrating to use discards. It's too easy to get trapped in a case where _ suddenly binds to an identifier and that will invalidate many other cases in the method where you depended on having discards available and there is little recourse for the developer when that happens.

I have a supposition we can fix that, without having to go whole-hog into: all `_` are always discards.

The open question for me is if there are cases where code would be legal under either interpretation (identifier or discard), and you want the latter, and interpretting as the former would lead to undesirable behavior.  If that exists, then this approach would likely not be viable.  However, my hunch is that this would allow for:

1. existing code to continue to compile with its existing meaning.
2. Code that is currently in error will now compile, with a meaning that is sensible.
3. Code that could potentially have both meanings (and this will retain the 'identifier' interpretation) will behave in a desirable way.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> Could we change that and instead make it so that if it's an error with the prior semantics, then it can now be reinterpretted as a discard?

I like where this is going but it sounds like there could be a lot of potentially tricky edge cases, especially if the code seems to intentionally mix discards and `_` as an identifier:

```cs
if (int.TryParse(s, out _)) {
    // ...
}
// later ...
var bar = foo.Select(_ => _.Bar)
```

**Comment by jaredpar** [comment]
> Could we change that and instead make it so that if it's an error with the prior semantics, then it can now be reinterpretted as a discard?

Can't do that because `_` is a legal identifier. As @HaloFour pointed out it's just fine to use it via `_.ToString()`, etc ... You can't even take shortcuts like saying "okay, if _ is only used for assignment or `out` then make it a discard" because assignments to a `_` can have side effects (implicit conversion tricks). 

This is the core problem we're facing. The decisions of C# 1.0 are essentially limiting our ability to make `_` a friction free feature. Unless we take some sort of conditional break here then we're essentially stuck with those decisions.

**Comment by CyrusNajmabadi** [comment]
> Can't do that because _ is a legal identifier. 

I'm not taking about the cases where is has legal, error free, semantics. 

I'm talking about the cases where it has illegal semantics. For example, where it would cause a scope collision. 

This code would be *illegal* today, and so we can come up with rules to make it legal by saying: ah, ready these all as discards now.

**Comment by jaredpar** [comment]
I'm unsure what you're asking for at this point.

**Comment by CyrusNajmabadi** [comment]
In cases where the code compiles today using current rules, preserve the meaning of that code. 

In cases where the code does not compile (for example, because of scope collision), allow reinterpretation as a discard. 

This is effectively similar to how other semantic identifiers work. If 'nameof' binds, then use that, otherwise it is the semantic keyword. 

Except instead of asking if it binds, ask if there is a scoping collision or not found, it things like that. In that case, treat as discard.

**Comment by 333fred** [comment]
That just sounds incredibly dangerous to me. Did you mean to use it as identifier and messed up, or did you actually want to discard? The intent of _ is to make the programmer's intent clear, but this would do the opposite.

**Comment by CyrusNajmabadi** [comment]
Like I said, I want to do the mental exercise here. 

With the other semantic identifiers the above holds (with the same arguments), but it really didn't turn it to be an issue. 

My supposition is that it will be very obvious very quickly.

**Comment by jaredpar** [comment]
Binding isn't enough though because it doesn't fix any of the problems I outlined. Today function parameters and single parameter lambdas are always bound as identifiers. Hence we can't take the approach we take with other contextual keywords (it's what we do today)

If we want to take the approach of "discard if it doesn't impact behavior" then that means we have to effectively implement two different binding passes. Because in order to determine if it's legal as discard, by that I mean doesn't change the side effects of the program, then you have to do semantic analysis. Have to understand for example if the current approach has silent implicit conversions. Consider the following as an example: 

```c#
M(x, _ => { _ = ""; });
```

In order to understand if it is legal to treat `_` as a discard or must be preserved as an identifier you must go through a full binding pass. It's completely possible that it binds to the following for which treating `_` as a discard in the future would be a breaking change. 

```
void M(Action<string, dynamic> a);
```

I don't think this is worth doing a double binding pass on methods which is why I'm pushing for other approaches.

**Comment by CyrusNajmabadi** [comment]
Here's why I don't think it's an issue. Say you run into this and we're now treating something as a discard that your didn't intend:

First, if you are actually using the thing as an identifier, you will find out practically immediately. That's because in practically any way you would use `_`, we would now error. I.e. `Console.WriteLine(_)`, or `=> _.ToString()`. So you would immediately get errors saying something was wrong, and it would be immediately after you added some new `_`. 

If you didn't get errors, then you really weren't using the identifier except to act as a discard anyways. So it read likely safe for us to reinterpret this way. 

What I do*not* know is: are there cases where your would do this where both interpretations are desirable, and we'd pick 'identifier', and that works be undesirable.

**Comment by CyrusNajmabadi** [comment]
> In order to understand if it is legal to treat _ as a discard or must be preserved as an identifier you must go through a full binding pass.

There is no scoping collision there.

**Comment by jaredpar** [comment]
To be clear then: you're okay with a solution that breaks compatibility in the case of implicit conversions? 

> There is no scoping collision there.

No but there is a breaking change for doing this. I can't tell from your arguments if you do or don't care about the breaking change aspect. I can read your arguments both ways.

**Comment by CyrusNajmabadi** [comment]
>  It's completely possible 

What I care about it what is probable and not necessarily pathological. What I'm trying to ascertain is not if we can construct some case, but whether such cases weighs be common or relevant.

We could construct ambiguous cases for many semantic keywords. Heck, that's a major discussion point around `field`. But the question is not if the case exists, but if changing meaning there is a significant issue for real code and users. My supposition is that it is not.

**Comment by CyrusNajmabadi** [comment]
> No but there is a breaking change for doing this

There should be no breaking changes with my position.  Because my position always has the rule: if this compiles legally with c#10 semantics, then it keeps that meaning. 

The open area of discussion for me is: are there*a subset* cases that do not compile that we could identify, make a sensible rule around that would unblock discards for a significant number of cases and users. 

I think it is worth investigating that avenue. It doesn't even feel strange or inappropriate for me to think about things in that manner. This is just how I see the semantic identifier space.

**Comment by jaredpar** [comment]
> There should be no breaking changes with my position. Because my position always has the rule: if this compiles legally with c#10 semantics, then it keeps that meaning.

Assignment can have side effects. As I've pointed out in examples above `_ = y` can introduce side effects and that is one of the **primary** uses of discards. In order to determine if it did or did not have side effects you must first do a full binding to analyze conversions.

**Comment by CyrusNajmabadi** [comment]
> I can't tell from your arguments if you do or don't care about the breaking change aspect.

I care deeply. A core design goal here for me is that if it compiled before, it will continue to compiler with the same meaning, with all else being equal (i.e., you didn't change references or something like that). 

That almost always is how I think of semantic identifiers, with `record` being a case where I was ok with not having that, as it would need a semantic type that collided there, and I'm ok stating the language will presume that that's not ok. For expr contexts I think we have to take a narrower view, but I think there may still be rules that can work. I think it's worth the exercise at least.

**Comment by CyrusNajmabadi** [comment]
> Assignment can have side effects. As I've pointed out in examples above _ = y can introduce side effects 

Why are we talking about that example though?  As I've said, of the example had legal semantics today, it wouldn't change. 

Your example didn't seem to fit into a car where I would expect us to reinterpret as a discard....

**Comment by CyrusNajmabadi** [comment]
I think my general point is not being heard. I'm not proposing a language rule here. I'm proposing a way of potentially exploring the space that I think a motivated working group *might* be able to use to expand the set of scenarios where discards would be allowed. 

It's a *supposition*. A theory that I think is worth exploring. It's not a claim that it definitively solves this. 

Furthermore, a core premise I have is that the new rules only get involved with errors (though which errors is TBD), similar to how we do things with 'var', 'nameof' etc. So compat is viewed as paramount. If I didn't care about compat, this space would be trivial :-)

**Comment by jaredpar** [comment]
> Your example didn't seem to fit into a car where I would expect us to reinterpret as a discard....

I posted an argument that went into great detail about how the fact that lambda parameters are treated as identifiers, not discards, is an enormous friction point. You have made arguments that we can detect these cases and allow them to be treated as discards. I have provide examples that your detection scheme will not work for straight forward examples without introducing multiple binding phases. That means that the `_` in the cases above will continue to be treated as an identifier and hence do nothing to solve the problems I've outlined. 

I think we need to take a step back at this point. The argument is not progressing. I've done my best to outline what I feel are the problems with discards today. You provided an idea and I feel like I've demonstrated how that isn't workable (without extra binding phases) due to compat issues. You don't seem to think I've done this hence we are talking past each other in some form. Think we should take a step back and try again later.

**Comment by CyrusNajmabadi** [comment]
I am literally not trying to solve all cases. As I have tried painstakingly to make clear, I am saying that I think there might be wishes here to improve things in ears for a subset of cases. It would be up to a working group to determine:

First, if that subset can be identified. 
Second, if that subset is suitably valuable. 

I have repeatedly stated that I think combat is important, and that I hope the above two hold even while preserving code meaning.

**Comment by CyrusNajmabadi** [comment]
> You have made arguments that we can detect these cases 

No. I have not. First, I am making no argument.  Second, I have not stated what cases I think weighs be solved here. I genuinely do not know, which is why I've suggested a working group to examine this. I have provided a framework I think might be valuable there, but I am fully under the belief that that working group might not find this approach to be successful (though I have a belief there is value here and success could be found).

**Comment by CyrusNajmabadi** [comment]
> You provided an idea and I feel like I've demonstrated how that isn't workable 

I don't see how. My idea doesn't even encompass your case. As that would be legal code, my compat portion of the design takes precedence. And, imo, that's sensible. If you provided me with that code, I*would* want it to preserve meaning. That's why I stated my views in the above posts about how in practical real world cases I feel like this will likely be fine. See https://github.com/dotnet/csharplang/issues/4460#issuecomment-813460646 for example.

**Comment by CyrusNajmabadi** [comment]
> without introducing multiple binding phases.

This is your personal complaint about a supposed possible rule that a wg might come up with. 

First, a wg might not come up with that rule. But say they did. Your position is totally fine, but would simply be an opinion into the ldm as to whether or not that was acceptable or not. 

For example, the `field` working group does something similar. We bind, and device what to do in the case of failure. It was also somewhat unpalatable to some. But we still felt we could create a rule there that was acceptable. 

Perhaps we are able to find an acceptable time here as well. It perhaps we come up with this rule and it is doing to be unacceptable.  That's all ok. What I'm not ok with us just preemptively disregarding the idea of this exploration becuse we feel it will only come up with this solution and this solution will necessarily be rejected.  It presupposes two sets of events which are not at all evident to me.

**Comment by alrz** [comment]
>this would allow for:
>
>1.existing code to continue to compile with its existing meaning.

Isn't this the philosophy behind the current semantics? You can make it *less* annoying by defining all kinds of rules but I think the point is to avoid that sort of stuff which we're already dealing with.

To me, any solution to this besides a (perhaps granual) breaking change doesnt make anyones life easier.

**Comment by CyrusNajmabadi** [comment]
> Isn't this the philosophy behind the current semantics?

Not as i understand it.  

For example, you can't write:

```
        void Foo(string _, int _)
        {
        }
```

even though that should be ok.

**Comment by CyrusNajmabadi** [comment]
> To me, any solution to this besides a (perhaps granual) breaking change doesnt make anyones life easier.

If we can make a breaking change, this gets simple :)  I'm totally ok with that, but i'm operating under the assumption that far too much code uses `_` in a sensible fashion to allow this.

**Comment by 333fred** [comment]
> > Isn't this the philosophy behind the current semantics?
> 
> Not as i understand it.
> 
> For example, you can't write:
> 
> ```
>         void Foo(string _, int _)
>         {
>         }
> ```
> 
> even though that should be ok.

Well, the way I interpret your proposal (and correct me if I'm wrong) is that either or both of those arguments could be discards. If you had a `string _` in scope, for example, the `string` would be an actual out param and the `int` would be a discard. That is what my objection was to earlier, as that seems insane to me. If that's not actually what you meant, then I'm afraid I'm not understanding what you meant :)

**Comment by CyrusNajmabadi** [comment]
> Well, the way I interpret your proposal (and correct me if I'm wrong) is that either or both of those arguments could be discards. 

As this code would not be legal (scope collision), they would be both be discards.

--

My genral point is: if there are any problems *all* are discards (i.e. they take on this new meaning).  It's not piecemeal where it's some `_` stay as identifiers and some become discards.

**Comment by 333fred** [comment]
> As this code would not be legal (scope collision), they would be both be discards.
> 
> --
> 
> My genral point is: if there are any problems _all_ are discards (i.e. they take on this new meaning). It's not piecemeal where it's some `_` stay as identifiers and some become discards.

I hate it :). That, imo, is just as confusing as the case where one is a discard and one binds to the name in scope. At least the current behavior is that that code doesn't compile. Suddenly, when I add or remove parameters I have to consider whether this change makes the code no longer compile by the old definition, which might cause the meaning of my code to change as it binds to or no longer binds to `_` identifiers in scope.

**Comment by CyrusNajmabadi** [comment]
>I hate it :). That, imo, is just as confusing as the case where one is a discard and one binds to the name in scope.

IMO, the end user effect is likely teh same.  I get to indicate "i don't care" in any of its forms.  If i do care (because i'm actually going to use it as an identifier and i read it), then it works.

**Comment by CyrusNajmabadi** [comment]
>  Suddenly, when I add or remove parameters I have to consider whether this change makes the code no longer compile by the old definition ... which might cause the meaning of my code to change as it binds to or no longer binds to _ identifiers in scope.

I would love us to do a real world test of this on real code to determine if this could actually ever really bite someone.

I'm not interested in pathological 'what ifs?'  I'm much more concerned with: what woudl happen in practice.  And, in reality, i think it would be staggeringly unlikely to happen.

You'd need the code to:

1. be using discards currently.
2. remove the use of a discard such that it now meant an identifier
3. be using the discard in a way in teh original code such that the new code still compiled as an identifier
4. have the meaning of that code change in a way that was now *noticible*.  i.e. since you could only write to the discard (you can't read from it) writing to it as an identifier woudl have to substantively change meaning as opposed to just dicarding.

Those all seem so enormously unlikely that i think we're just attempting to not even give this a fair shot.

I'll also note that such a possibility already exists with discards today *and we accepted it*.  Say i have `Foo((_, _) => ...)` (haha, looks like a butt).  Well, we allow that today, *and* we have the aspect of this that if you remove a discard it becomes an identifier.  Sure... maybe that somehow breaks someone who didn't expect that behavior.  But gosh if that just doesn't seem like an incredibly unlikely or narrow thing to care about.

I'd much rather focus on normal code doing normal things, the vast majority would continue to operate sanely IMO even with this sort of mentality toward discard.

**Comment by 333fred** [comment]
> > Suddenly, when I add or remove parameters I have to consider whether this change makes the code no longer compile by the old definition ... which might cause the meaning of my code to change as it binds to or no longer binds to _ identifiers in scope.
> 
> I would love us to do a real world test of this on real code to determine if this could actually ever really bite someone.
> 
> I'm not interested in pathological 'what ifs?' I'm much more concerned with: what woudl happen in practice. And, in reality, i think it would be staggeringly unlikely to happen.
> 
> You'd need the code to:
> 
>     1. be using discards currently.
> 
>     2. remove the use of a discard such that it now meant an identifier
> 
>     3. be using the discard in a way in teh original code such that the new code still compiled as an identifier
> 
>     4. have the meaning of that code change in a way that was now _noticible_.  i.e. since you could only write to the discard (you can't read from it) writing to it as an identifier woudl have to substantively change meaning as opposed to just dicarding.
> 
> 
> Those all seem so enormously unlikely that i think we're just attempting to not even give this a fair shot.

I would agree that those are pretty unlikely to happen, but I also think that, other than lambda parameters named `_` for ease of typing that one parameter, the identifier `_` doesn't occur much in the wild. I'd love to see us try and ban it outright :).

> I'll also note that such a possibility already exists with discards today _and we accepted it_. Say i have `Foo((_, _) => ...)` (haha, looks like a butt). Well, we allow that today, _and_ we have the aspect of this that if you remove a discard it becomes an identifier. Sure... maybe that somehow breaks someone who didn't expect that behavior. But gosh if that just doesn't seem like an incredibly unlikely or narrow thing to care about.

This case seems much worse to me than the double underscore lambda case. That happens at the introduction of the names. and messing it up generally causes code to no longer compile. What you're proposing does the opposite: it makes code that _didn't_ compile now compile. With the exception of definite assignment, I'm struggling to find a case where the "I did the wrong thing" would manifest as a compile error, rather than insidious runtime bug.

**Comment by CyrusNajmabadi** [comment]
>  I'd love to see us try and ban it outright :).

Well, like i said, if we can just break code, then this entire discussion is moot. My theory was predicated on the: us not being able to ban an acceptable coding pattern that we know does exist.  If we have a path to avoid that, then by all means i'm fine with that :)

**Comment by CyrusNajmabadi** [comment]
> rather than insidious runtime bug.

I'd love a non pathological "insidious runtime bug" case.  Even the pathological things i can think of seem pretty silly.  Liking need a implicit conversion to run that also has side effects.  And if your implicit conversions have side effects *And* you were assignign to something called `_`... well, then i'm ok with breaking that.

**Comment by HaloFour** [comment]
It's a shame that `_ => _` caught on in lambda bodies as a form of method reference.  If it weren't for that the likelihood of an `_` being used as an actual identifier would probably be a lot closer to non-existent and the concerns about treating it as a discard in all cases would largely go away.  I'd be curious as to how prevalent `_ => _` actually is in the wild.

**Comment by 333fred** [comment]
> I'd love a non pathological "insidious runtime bug" case :). Even the pathological things i can think of seem pretty silly. Liking need a implicit conversion to run that also has side effects. And if your implicit conversions have side effects _And_ you were assignign to something called `_`... well, then i'm ok with breaking that.

I don't think anything particularly terrible is needed, just assume that your `_` variable wasn't assigned to and you were looking at the value in it later.

**Comment by CyrusNajmabadi** [comment]
>  and you were looking at the value in it later.

If you're looking at the value in it later... then that could only ever be a non-discard (as discard values can't be read from).  So i'm not seeing the problem.  There's no interpretation in which treating as a discard would be legal.

**Comment by CyrusNajmabadi** [comment]
> If it weren't for that the likelihood of an _ being used as an actual identifier

It's also used for `using var _ = ...`  Though, if we just retcon that to be a discard as well (using the logic i outlined above), then we'll be ok.  I would actually be heavily in support of that as i want to be able to write:

```
using var _ = ArrayBuilder<X>.GetInstance(out var xs);
using var _ = ArrayBuilder<Y>.GetInstance(out var ys);
```

Right now IDE has to instead do:

```
using var _1 = ArrayBuilder<X>.GetInstance(out var xs);
using var _2 = ArrayBuilder<Y>.GetInstance(out var ys);
```

Which is annoying.

**Comment by CyrusNajmabadi** [comment]
The above is an example of something that i think woudl nicely fall out from the description i gave.  As there are two `_`s in scope, it couldn't ever legally be an identifier.  So it would just be a discard you could yourself not reference (but compiler could still do things like `using` for.

**Comment by 333fred** [comment]
> > and you were looking at the value in it later.
> 
> If you're looking at the value in it later... then that could only ever be a non-discard (as discard values can't be read from). So i'm not seeing the problem. There's no interpretation in which treating as a discard would be legal.

```cs
int _ = 1;
MyMethod(out _ /* int */, out _ /* string */); // As I understand your proposal, both of these would be discards
if (_ == 1) { DoSomething(); }
```

Obviously the real code would have to be more complicated than this to not be obviously broken, but this is what I'm understanding from your proposal.

**Comment by CyrusNajmabadi** [comment]
> // As I understand your proposal, both of these would be discards

No.  In my proposal *all* of these would be discards.  The code isn't legal today, so we can take on the new semantics.  IN the new semantics, all the `_` in that method are discards.

It's somewhat akin to how `async` works.  `await` is unilaterally a keyword inside of it.  In my world *except* for preserving back compat, we then take the meaning that `_` is now *ours* and has this meaning.  You don't get to mix/match.

**Comment by 333fred** [comment]
> n my proposal _all_ of these would be discards :).

I was worried we were talking past each other, and indeed we were. I am much more amenable to your proposal now, and I don't think it would necessarily need a separate binding pass. Seems like it should be possible to do in one.

**Comment by CyrusNajmabadi** [comment]
Sorry about that.  It's not a fully formed proposal in my head.  So i'ts more about taking a few incomplete threads and trying to describe them and how they all work together.  Easy to have gaps there.

**Comment by jaredpar** [comment]
I do not understand at all what is being proposed. I have taken the time to outline in detail what I think is wrong today, what I think could be wrong about what is being proposed and I'm just not getting feedback on that.

**Comment by CyrusNajmabadi** [comment]
What is being proposed is a working group to examine the idea of detecting situations where we would switch to a mode where all _ in a method would be treated as discards. The proposal asks the working group to define the scenarios in which this would happen and show:

1. What currently illegal code would be made legal. 
2. How to ensure that currently legal code retains its meaning.
3. What sort of situations might lead to confusion. For example what code might make people think discards were in use when it was actually an identifier. 
4. A determination of the confusing cases would be likely or not, and in either case if it would be likely that this would actually lead to problems in practice. 
5.

**Comment by jaredpar** [comment]
1 + 2 together are incompatible goals. Consider this as a concrete case:

```c#
void M(dynamic d, string _) {
  _ = d;
}
```

Goal 1 says `_` should be a discard at which point this is a discarded where as today it's a dynamic conversion operation. This is the implicit conversion problem I keep referring to. It's an unavoidable issue when you more broadly treat `_` as discards within a method body and extends beyond `dynamic`. 

I don't think this is a bad break to change, it's actually quite acceptable to me. At the same time I think it needs to be acknowledged. Early proposals by me stating we would need breaking changes were pushed back on because I was up front about this being an unavoidable side effect. This proposal is no different. It's an explicit call for breaking changes, it should be upfront about it.

**Comment by jmarolf** [comment]
I am fine pursuing a separate proposal that outlines what it would take to smooth out issues with `_`. I do see a lot of places where folks want "pseudo-discards"

like this pattern:
```csharp
using var _ = ArrayBuilder<X>.GetInstance(out var xs);
```
Of the lambda one:
```csharp
Func<int, int> action2 = _=>_;
```

for all the other keywords I am confident saying that there is no benefit to keeping them as identifiers in existing code, its only going to confuse people who look at it laters. But discards cannot do several things today and so folks rely on the fact that they can be identifiers as well to work around this.

For @CyrusNajmabadi's pattern, if you were just allowed to dispose a discard like:

```csharp
using _ = ArrayBuilder<X>.GetInstance(out var xs);
```
or

```csharp
using (_ = ArrayBuilder<X>.GetInstance(out var xs))
{
...
}
```

I then think we could move forward. Basically my understanding is:

1. identify all the pseudo-discard patterns that folks might be relying on today
2. Have an answer for them in the discard feature itself
3. move to make `_` a lexical keyword

**Comment by CyrusNajmabadi** [comment]
> Goal 1 says _ should be a discard at which point this is a discarded where as today it's a dynamic conversion operation. This is the implicit conversion problem I keep referring to. 

This is legal code.  So would remain legal under teh approach i'm supposing.


> Goal 1 says _ should be a discard 

First: No it doesn't.  You're misreading goal 1.  Goal 1 is not: all illegal code becomes legal.  Goal 1 is: identify illegal code that could be made legal.

Second: This is legal code.  So, as i've stated numerous times, i would expect it to preserve meaning.  I feel like i've been consistent stating this.

> I don't think this is a bad break to change

That's fine.  I have never stated that i'm opposed to a breaking change approach.  I have only stated that i think there is merit in investigating a non-breaking change approach.  I am not arguing against any other potential approaches here.  I'm just offering my thoughts on something i think might be viable.

**Comment by CyrusNajmabadi** [comment]
> or
> ```c#
> using (_ = ArrayBuilder<X>.GetInstance(out var xs))
> {
> ...
> }

This would defeat the purpose.  We'd just write:

```c#
using (ArrayBuilder<X>.GetInstance(out var xs))
{
...
}
```

for that :)

The goal is to be able to use `using locals` instead of `using statement` so we can avoid the block :)  We coudl call this a `using dicard`.

**Comment by CyrusNajmabadi** [comment]
>  There are patterns today with psedo-discards that are solving real problems for people and we should at least know about them before making a call on this.

100% :)  I don't think it would be hard to identify.  We could likely discover practically all of them just by analyzing all teh repos at our disposal.

**Comment by 333fred** [comment]
> > There are patterns today with psedo-discards that are solving real problems for people and we should at least know about them before making a call on this.
> 
> 100% :) I don't think it would be hard to identify. We could likely discover practically all of them just by analyzing all teh repos at our disposal.

Here's one example I remember seeing off the top of my head (which uses `_` as a method name, something none of us have discussed until now): https://github.com/morelinq/MoreLINQ/blob/master/MoreLinq/Choose.cs#L58-L67

**Comment by CyrusNajmabadi** [comment]
> Here's one example I remember seeing off the top of my head (which uses _ as a method name, something none of us have discussed until now): https://github.com/morelinq/MoreLINQ/blob/master/MoreLinq/Choose.cs#L58-L67

I...

Uh...

:'(

/sob

**Comment by Youssef1313** [comment]
> Here's one example I remember seeing off the top of my head (which uses `_` as a method name, something none of us have discussed until now): https://github.com/morelinq/MoreLINQ/blob/master/MoreLinq/Choose.cs#L58-L67

Well, also here is a class name `_` in dotnet/reactive :smile:

https://github.com/dotnet/reactive/blob/5753528bd710a2e7a71000d7a63f777e58d4e591/Rx.NET/Source/src/System.Reactive/Linq/Observable/Latest.cs#L22

**Comment by HaloFour** [comment]
Kill it with 🔥.

**Comment by jmarolf** [comment]
So in these two examples:

- [local function named `_`](https://github.com/morelinq/MoreLINQ/blob/ee238241083a9b31dc03c39781b29f4189b4fe73/MoreLinq/Choose.cs#L58-L67)
- [private class named `_`](https://github.com/dotnet/reactive/blob/2c022ef9386432ead0829f462d3910772da3a790/Rx.NET/Source/src/System.Reactive/Linq/Observable/Latest.cs#L22)

I feel _very_ comfortable breaking these cases. I will admit I take the axiom of meaning to be something that should always be held (or at least attempted) in code. Attempt to give meaningful names to variables, functions, types etc. In these cases the use of `_` appears to be an abdication of naming responsibilities. All of these are private and there is no need to use `_` to communicate that these members/types should not be used. If a future update told the maintainers of this code that they _at most_ must rename these to `@_`.

**Comment by CyrusNajmabadi** [comment]
In the idea i have, these would continue to be fine.  A discard couldn't ever appear in places where you would use these identifiers (which would practically always be reads, and i believe always be syntactically apparent as such).  

Honestly, i don't see these as things that actually block us from just moving to discard interpretation in more cases.

-

Let me walk through why i think this approach is ok, and also totally complimentary to how C# has done things up to now.  Let's use the lambda case as where we might want discards to now be inferred, and why it would be ok to start doing that.

Let's recall how we added lambdas to the language all the way back in C# 3.0, first through an example:

```c#
using X;
using Y;

...
x.Select(y => y.M(0));
...
```

First, the way this will work in the language is that we will collect all the potential `Select` members that could be invoked, and we'll then attempt to interpret `y => y.M()` against the parameter in `Select` that that argument corresponds to.  We will eliminate any of these interpretations that result in an error, and only succeed if there is exactly one interpretation that succeeds.

Now, let's think about all the (non-exhaustive) ways that interpretation could affect the meaning of that code:

1. Perhaps that is a lambda.  Perhaps it is an expression tree.
2. Perhaps M is actually invoked, perhaps it is not (For example, if it has a Conditional attribute on it).
3. Perhaps this is a static invocation, perhaps it is an instance invocation (the `Color Color` case)
4. Perhaps `y` has one of many different potential types.
5. Perhaps this is a Func, perhaps it is an Action
6. Perhaps this is invoking a method, perhaps it is invoking a member with delegate type.
7. Perhaps this is an instance method.  Perhaps it is an extension method.
8. Perhaps this is a generic invocation with inferred type parameters, perhaps it is not.
9. Under different interpretations a different set of `using` directives may be 'used'.  Even crazier, we may pick an interpretation that both requires that a `using` be in teh code to ensure its meaning, even if the interpretation we chose uses *nothing* from taht `using` directive.  This happens when a `using` directive causes *another* interpretation to `error out`.  We must preserve that `using` as removing it would cause multiple successful interpretations, which would be an error.
10. etc. etc.

As we can see, it is *already* teh case that one cannot syntactically decipher the meaning of a piece of code, even when it appears utterly obvious what it must mean.  It has always been the case that to really understand, you need to understand *all* surrounding code and *all* binding.  That's why tooling is often so critical.  It allows a way for this information to light up to the user simply to answer all those potential confusing interpretation concerns.

As such, i see no problem, and i personally have no concerns with any approaches that involve potentially binding several times, and taking on new views of the code depending on that additional binding interpretation.  Yes, it would be something people might occasionally need to be aware of.  But that seems enormously minor in the grand scheme of things given the wide set of *basic* cases that already exist that people *already* need to be very aware of today.

Furthermore, i posit that in practice, just as the above is likely not to be too confusing, the discard case (even with rebinding) would not be confusing.  I *fully* agree that pathological cases may exist (just as they do for all the cases i enumerated above).  But that is not what it think is relevant.  What i think is relevant is if those pathological experiences will genuinely arise naturally, and even if so, if they woudl be common, or highly unlikely to occur.  I believe it's likely to not naturally happen, and thus overly fretting about it *especially* in a world where the above cases occur is unnecessary.

Finally, i'll point out that for all the cases above where we do have the problem of syntax not being clear enough.  We've always been willing ot lean on the idea that users can access rich tools to address their concerns and confusion.  I do not see why that would not be true in a future where we support discards in more places.  In a world where the majority of our users can use VS, VSCode, Rider, Resharper, or an LSP host that can use an appropraitely powered LSP server, this should not be confusing.  The tooling will show these as discards when tehy are, and will not when they're not.  Will there be locations without this data?  Yes.  But we've seen that users are able to make due when the same situation arises with the same standard language features we've been shipping since 1.0.  So i do not feel that discards are new or special in this regard.  And i do not accept this issue as a necessary blocker of this approach.

Thanks! :)

**Comment by jmarolf** [comment]
Generally agree with this on principle. I'll admit I am a bit lost as to the specifics @CyrusNajmabadi could you write a formal proposal so I can see in greater detail what you are thinking? 

> Finally, i'll point out that for all the cases above where we do have the problem of syntax not being clear enough. We've always been willing ot lean on the idea that users can access rich tools to address their concerns and confusion. I do not see why that would not be true in a future where we support discards in more places. 

I think this needs to be weighed against the inherent confusion this feature creates. While we should always desire to have tooling enhance the clarity of language constructs, I do not think it can fix things that are inherently broken. 

If we feel that `_` should be kept as a valid identifier in perpetuity I would propose that we add warnings to a future warning wave so that programmers can state that using `_` as a name is their explicit intent instead of them just misunderstanding/misuse of the feature.

**Comment by Korporal** [comment]
@jmarolf - If I understand this correctly the objective is to eliminate reserved words, yes? Some language grammars have this ability, _no words_ were reserved, an identifier _could_ be the same as a keyword without difficulty. The reason for this was to allow the future addition of new keywords without breaking existing code.

However C# like Java, C++ and C before it have a basic grammar that makes this impossible and I don't think this could be changed without - to all intents and purpose - defining a new language.

Having no reserved words made future enhancements to the language pretty easy but this was a goal from day one and unless a grammar is designed with that in mind I don't think it can ever be added later.

In PL/I parlance _all_ statements in the language were either

`<reference> = <expression>` (i.e. an assignment)

or

`<keyword-statement>`

The parser would attempt to parse each statement as the former and if that failed then it would attempt to parse it as the latter.

So if one had code like this:

`start = 100 + T;`

and the language designers later wanted to add a new keyword called `start` they could, because the parser is able to distinguish between these scenarios.

A common demo of how this worked was something like this, which compiled and ran without any problems:

```
/* This is not advocated as actual code but is entirely valid */
if if = else then
   else = then 
else
   then = if;
```

In this sense every keyword was contextual because there were only two possible contexts as I mentioned above.

The above can be broken down as:

```
if <expression> then
   <assignment>
else
   <assignment>
```

**Comment by CyrusNajmabadi** [comment]
@Korporal that wouldn't work here as the pl1 grammar is too limited and too verbose for too many constructs we care about.

**Comment by Korporal** [comment]
Mr. Najmabadi, the C# grammar is "basic" in the sense that it has reserved words (as does Java, C++ etc.) the absence of which - in _some_ languages - is a potent feature, as I was discussing in my remarks above. I have no desire to enter into a pointless back and forth over this either.

**Comment by CyrusNajmabadi** [comment]
@Korporal you brought the PL family of languages into the discussion :)  My point was simply that the rules the employ to allow for this aspect of that family are too restrictive for modern languages.  It would feel artificially constraining and would force us to avoid many language constructs that we would want.

**Comment by CyrusNajmabadi** [comment]
@Korporal I've hidden your comment for being disruptive (and Halo's for being off topic).  Please try to follow the .net CoC listed here: https://dotnetfoundation.org/about/code-of-conduct

There was no trolling here.  These are my professional opinions on the topic.  Please try to be respectful of differing opinions and  viewpoints.  Thank you.

**Comment by jmarolf** [comment]
> If I understand this correctly the objective is to eliminate reserved words, yes? 

@Korporal The goal of this proposal is the opposite. imho C# is too lenient allowing developers to define types such as `var`, `dynamic`, `record`, or `_`. There are some good reasons for these lookup rules (as you've noted) because they allow for maximum backward compatibility. However, my argument is that this makes the language needlessly complex. I do not want a developer to see `var` or `dynamic` and ever need to wonder if that is the keyword or a user-defined type.

**Comment by Korporal** [comment]
@jmarolf - I'd like to see a list of all (lets say as of C# 10.0) `reserved` words _and_ `keywords`, two lists. Obviously these lists are not identical.

But you do make a good point I think and that seems to be (if I correctly follow you) that all keywords should - ideally - also become reserved words.

But then as you say, we hit the backward compatibility problem and that's a serious concern, I don't see anyway around this.

As keywords get added (and why shouldn't they as the language grows?) the list that a developer must be cognizant of also increases and that's a burden for them.

The point I was trying to make is that unless a grammar inherently avoids reserved words this nasty problem will emerge and is not easily solved, yes its far too late in the case of C# but not for say D#...

**Comment by CyrusNajmabadi** [comment]
>  I'd like to see a list of all (lets say as of C# 10.0) reserved words and keywords, two lists. Obviously these lists are not identical.

Can you clarify what you mean by 'rserved word' vs 'keyword'.  Keywords, definitionally (https://github.com/dotnet/csharplang/blob/main/spec/lexical-structure.md#keywords) are already 'reserved'.  We also have some more syntactic constructs, which are only legal to parse in a particular fashion, that can use an term which can be an identifier in other contexts.  

This is the list:

```
add
alias
and
ascending
assembly
async
await
by
descending
equals
field
from
get
global
group
init
into
join
let
method
module
not
on
or
orderby
param
partial
property
record
remove
select
set
type
typevar
when
where
with
yield
```

There are also special semantic identifiers.  But those are in an entirely different category, and can be though of more as builtins that the compiler consults in either the start or end of certain searches.

**Comment by Korporal** [comment]
@tannergooding - Thanks that is exactly what I was looking for. I have a question then now, are there any existing C# keywords that could be made into contextual keywords just by a change to the parsing logic? or are all of them unavoidably reserved as a result of the grammar?

**Comment by Korporal** [comment]
Hmm, yes I see these are good examples, so it is the grammar that leads to this not a parser implementation.

**Comment by CyrusNajmabadi** [comment]
> So are contextual keywords viewed favorably or not?

They are a language feature with pros and cons (listed in depth above).  Whether or not you view those pros/cons in a net favorable or net unfavorable way depends on how you view those pros/cons.  

> but its deemed undesirable to have more contextual keywords, 

Contextual keywords are not deemed as undesirable.  They are just a tool that come with a set of trade-offs, and the costs of those tradeoffs need to be assessed against the benefits.

Note that this calculation is not fixed in time, but changes as the language develops and grows larger.

**Comment by Korporal** [comment]
I'm curious as to what a grammar would, could look like that strived to be as close as possible to the current C# syntax yet was changed just enough to eliminate reserved words. This is something that interests me from an academic point of view.

I can envisage such a grammar being readily translatable into C# - as a proof of concept.

Some basic changes would be to use a keyword "call" (say) to invoke void methods and addition of "then" on an "if" to remove the ambiguity you pointed out above (is this a branch statement or an invocation of a method named "if"...).

I have no real idea though of the scale of something like that and as you say, this is deviating far from the intent of these discussions.

Thanks

**Comment by Korporal** [comment]
Well I'd really need a more substantial example of the `if (x) then y` scenario you're alluding to, in order to form an opinion. I guess I'm also still not really clear on how having no reserved words can be regarded as a limitation when it enables one to invent new keywords with as much complexity as is needed, without ever breaking existing code.

You say "if people want to use keywords today, they already can" yet you seem to be missing the point, it is not a desire to allow people to use keywords for identifiers but for new keywords to be introduced to the language without breaking existing code that may have (unknowingly) use words that we now want to use as keywords, maybe I had not made that clear.

C# recently introduced `record` and that is a contextual keyword - you were able to add it despite the fact that some code somewhere might have used `record` as a name, and this is a great ability I am simply arguing that all such terms be contextual ideally.

I can use `record` as an identifier but not `class` or `struct` 

Legal:

```cs
    public class record
    {

    }
```

Illegal

```cs
    public record class
    {

    }
```

If "more things were keywords" then the legal example above would never be allowed - yet might already exist and predate any introduction of `record` to the language.

Maybe I'm misunderstanding you...

**Comment by Korporal** [comment]
> Is there a code cleanup option in VS that will automatically add '@' in front of any identifier with a name that matches a contextual keyword in the language? Seems to me it would make it easier to address the backcompat issue if developers could turn on an analyzer that identifies anywhere that 'var' or 'async' or 'record' or '_' etc are used as identifiers, and then fix all those instances in a single click across their entire codebase. That would pave the way for turning that analyzer into a warning in some later version, and then making them real keywords eventually.

I don't know of such a thing and it seems to raise the problem that the code would have to already compile cleanly for such a tool to succeed yet how could it if it were using keywords invalidly?

**Comment by CyrusNajmabadi** [comment]
> but for new keywords to be introduced to the language without ever breaking existing code that may have (unknowingly) used words that we now want to use as keywords, maybe I had not made that clear.

**Comment by CyrusNajmabadi** [comment]
> C# recently introduced record and that is a contextual keyword - you were able to add it despite the fact that some code somewhere might have used record as a name, 

We were not.  We made a breaking change to support this.  

> and this is a great ability I am simply arguing that all such terms be contextual ideally.

This is not ideal, for all the reasons i outlined above.  Contextual keywords are a mechanism that can allow introduction, but come with drawbacks (again, listed numerous times above).

**Comment by Korporal** [comment]
> > yet you seem to be missing the point, it is not a desire to allow people to use keywords for identifiers but for new keywords to be introduced to the language
> 
> I'm not missing the point. You've talked about two separate things. In the words quoted here you're talking about new keywords being introduced. But in your immediately prior post you wrote:
> 
> > I'm curious as to what a grammar would, could look like that strived to be as close as possible to the current C# syntax yet was changed just enough to eliminate reserved words. This is something that interests me from an academic point of view.
> 
> You literally asked about eliminating reserved words, so i was discussing eliminating reserved words. if you want to discuss _adding_ keywords, that's also fine. But these are two different topics, and my responses about one of your questions should not be dismissed because you also want to discuss something else.

By eliminating reserved words Cyrus I meant eliminating them as a grammatical concept in a language by changing the grammar, and the motive for that is to then enable new keywords to be added to a language without ever breaking existing code because adding them can no longer lead to conflicts because reserved words are a grammatical impossibility.

If I failed to make that clear then I apologize.

**Comment by jmarolf** [comment]
> So are contextual keywords viewed favorably or not?

@Korporal from a language grammar perspective I have no opinion. They are certainly useful to allow for language evolution without breaking existing code. 

This proposal attempts to argue that the way contextual keywords were applied to a specific language (C#) make it more complex to use. When language concepts such as `nameof` can also be a method invocation it becomes another "gotcha" for new (and even experienced) programmers. However, even this proposal is not arguing that all contextual keywords be removed. Only those that require naming lookup rules. Contextual keywords that are never syntactically ambiguous (for example `record`) are fine to stay the way they are.

Related to this topic is the discussion of warning users about the user of all-lower-case type names:
- https://github.com/dotnet/roslyn/issues/56653
- https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-13.md#warning-on-lowercase-type-names

**Comment by Korporal** [comment]
> > So are contextual keywords viewed favorably or not?
> 
> @Korporal from a language grammar perspective I have no opinion. They are certainly useful to allow for language evolution without breaking existing code.
> 
> This proposal attempts to argue that the way contextual keywords were applied to a specific language (C#) make it more complex to use. When language concepts such as `nameof` can also be a method invocation it becomes another "gotcha" for new (and even experienced) programmers. However, even this proposal is not arguing that all contextual keywords be removed. Only those that require naming lookup rules. Contextual keywords that are never syntactically ambiguous (for example `record`) are fine to stay the way they are.
> 
> Related to this topic is the discussion of warning users about the user of all-lower-case type names:
> 
> * [Compiler or analyzer should warn for lower-cased type names roslyn#56653](https://github.com/dotnet/roslyn/issues/56653)
> * https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-13.md#warning-on-lowercase-type-names

You say

"Contextual keywords that are never syntactically ambiguous" but surely such situations would never compile, if the compiler and grammar can't resolve an apparent ambiguity then its a syntax error, no?

**Comment by CyrusNajmabadi** [comment]
> To a novice with C#

Yes. This is one of the reasons we've approved making a warning for this. So we can tell users: this likely is a bad idea, and is likely very confusing. 

And this is just one case. If there were no keywords, then you get this with *everything*.

**Comment by Korporal** [comment]
Well I could respond and counter some of the issues you're raising and misunderstandings that you seem to have above, but if I do so you'll likely start to complain that you dislike the 'direction' _our_ conversation has taken despite the fact that _your_ provocative responses directly contribute to that 'direction'. 

You make many bold assertions yet when challenged cry foul that I am "going in circles" (I am not) or "there's nothing to be gained" (there might be). 

Here are some of the assertions you've made recently:

> It cannot be both close to the current C# syntax, and yet not support even basic constructs like this.

> it would be such a departure as to not even feel slightly close to c#

> As such, such an approach would not be viable or desirable

> This is a degradation in user experience.

> We don't have an issue introducing syntactic keywords

> it's an anti-goal

>  it's a net negative for usability.

>  If there were no keywords, then you get this with everything.

>  It would certainly be confusing, for example, if if( might now be an invocation

Each of these can be challenged, some are opinions, some of them betray a misunderstanding of what I've said and some are simply plain wrong and others are a mischaracterization of things I may have said.

You feel it appropriate that you be permitted to make all kinds of claims yet inappropriate for me to disagree with you and when I do you then complain that I've taken the conversation in a direction you disapprove of. 

Why not stop responding to me altogether, if you did that and I was free to interact with the others here without being harassed then you'd see that I've actually made just three posts:

https://github.com/dotnet/csharplang/issues/4460#issuecomment-945196137

https://github.com/dotnet/csharplang/issues/4460#issuecomment-947123940

https://github.com/dotnet/csharplang/issues/4460#issuecomment-947693199

**Comment by CyrusNajmabadi** [comment]
@Korporal, as specified above, please open a new discussion, or take this tangent to gitter and/or discord.   As i said before i made a mistake here, believing this to be a discussion already and not an issue.  That was my fault and I apologize for it.  However, from this point on, any further conversation on this tangent is not appropriate for this issue.

**Comment by jmarolf** [comment]
> Why not stop responding to me altogether,

lol! That is not the @CyrusNajmabadi way :)

> "Contextual keywords that are never syntactically ambiguous" but surely such situations would never compile, if the compiler and grammar can't resolve an apparent ambiguity then its a syntax error, no?

@Korporal Sorry I was wrong here. I should not have said "syntactically ambiguous". As you say, it will parse to _something_. The question is whether that something is useful. 

The crux of this issue is that this is valid code in C# today:

```csharp
class await
{
    async async async(async async) => await async[async][async][async][async];
}
```
<details>
<summary>full code sample</summary>

```csharp
using System;
using System.Runtime.CompilerServices;

[AsyncMethodBuilder(typeof(builder))]
class async {
    public async this[async async] { get => null; }
}

class await
{
    async async async(async async) => await async[async][async][async][async];
}

static class util { 
   public static awaiter GetAwaiter(this await a) => throw null;
   public static awaiter GetAwaiter(this async a) => throw null;
}

class awaiter : INotifyCompletion {
    public bool IsCompleted => true;
    public void GetResult() { }
    public void OnCompleted(Action continuation) { }
}

class builder
{
    public builder() { }
    public static builder Create() => throw null;
    public void SetResult() { }
    public void SetException(Exception e) { }
    public void Start<TStateMachine>(ref TStateMachine stateMachine)
        where TStateMachine : IAsyncStateMachine => throw null;
    public async Task => throw null;
    public void AwaitOnCompleted<TAwaiter, TStateMachine>(
        ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : INotifyCompletion
        where TStateMachine : IAsyncStateMachine => throw null;
    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(
        ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : ICriticalNotifyCompletion
        where TStateMachine : IAsyncStateMachine => throw null;
    p
```

</details>

While this is a technical marvel from a parsing perspective, I ask the question "Does this make the language easier to use?"

This proposal argues that is does not. If you read that sample above and say "That seems fine, or at least not harmful" then you certainly don't agree with this proposal. If you also want more language constructs to behave this way (which is what I am interpreting your statements as being about) I agree with @CyrusNajmabadi that we should have that discussion on a separate issue. I am happy to start that if you want.

**Comment by TahirAhmadov** [comment]
I would say, let's keep `_` as contextual, but the other keywords mentioned should be prohibited from being identifiers. I don't have a single identifier in all of my codebases named `var` etc.; it's anecdotal evidence but I suspect there aren't many, if any at all, in other teams' projects either. And IDE can probably suggest a fix for these together with the release of this breaking change, making it a trivial 15-minutes exercise of fixing these build errors, using rename refactors. On the other hand, I do use `_` from time to time. Imagine method calls like `Info.Of<Person>().Property(_=>_.FirstName)` - in all of these cases, I purposefully use `_` to signify that it's not a real parameter and this whole construct is just a workaround for the time being, until something like `infoof` or `propertyof` is added.
I'm sure some or all of these points have already been covered - just don't have the time to read the entire thread.

**Comment by Joe4evr** [comment]
> it's anecdotal evidence but I suspect there aren't many

I agree there likely aren't many, but from what I understand, there are *some* big companies and/or government orgs whose code style policies are such that they never want to have `var` or the like, so they define some `public struct var/dynamic { }` to make the compiler issue an error since then any use will bind to the user-defined type instead of as a keyword. ***In the past*** it could've been quite costly for such a code base to be broken by a compiler/SDK update that started disallowing the use of such keywords as typenames, so even if there aren't many, those kinds of customers are worth *a lot* of money, and I have my doubts that Microsoft would've liked to risk losing them over such cases.

But today there are much better ways to enforce such a policy: There's a whole section of EditorConfig rules for it, so there's a clear migration path for such code bases that wish to ban the use of certain keywords in their entirety *while also* (possibly?) making it easier for the LDT to add new useful keywords in the future. (That said, the EditorConfig section in question is only for `var` at this time. It'd probably be good to have similar rules for `dynamic` and others, but that's a different discussion.)
